tendrl/gluster_integration/sds_sync/__init__.py

```diff
@@ -930,3 +924,91 @@ def get_volume_alert_counts():
+                # find ancestors if mountpoint present
+                if dev_info["MOUNTPOINT"]:
+                    ancestor_details = find_ancestors(
+                        block_devices, dev_info["KNAME"]
+                    )
+                    if dev_info["MOUNTPOINT"] in devicetree.keys():
+                        # if different block device mounted in same mountpoint
+                        # then append the disk and partition information under
+                        # same mountpoint
+                        disks = devicetree[dev_info["MOUNTPOINT"]]["disks"]
+                        devicetree[dev_info["MOUNTPOINT"]]["disks"] = list(
+                            set(disks + ancestor_details["disks"])
+                        )
+                        devicetree[dev_info["MOUNTPOINT"]]["partitions"] = \
+                            list(
+                                set(ancestor_details["partitions"])
+                            )
+                    else:
+                        devicetree[dev_info["MOUNTPOINT"]] = {}
+                        devicetree[dev_info["MOUNTPOINT"]]["disks"] = \
+                            ancestor_details["disks"]
+                        devicetree[dev_info["MOUNTPOINT"]]["partitions"] = \
+                            ancestor_details["partitions"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"] = {}
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "size"] = dev_info["SIZE"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "mountpoint"] = dev_info["MOUNTPOINT"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "type"] = dev_info["TYPE"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "size"] = dev_info["SIZE"]
+        return devicetree
+    except Exception as ex:
+        logger.log(
+            "error",
+            NS.publisher_id,
+            {"message": "Unable to generate devicetree.err: %s" % ex}
+        )
+
+
+def find_ancestors(block_devices, device_name):
+    # find ancestors including current dev from bottom to top
+    ancestor_details = {}
+    ancestor_details["disks"] = []
+    ancestor_details["partitions"] = []
+    while True:
+        if device_name in block_devices.keys():
+            if block_devices[device_name]["TYPE"] == "disk":
+                ancestor_details["disks"].append(
+                    block_devices[device_name]["NAME"]
+                )
+                break
+            elif block_devices[device_name]["TYPE"] == "part":
+                ancestor_details["partitions"].append(
+                    block_devices[device_name]["NAME"]
+                )
+                device_name = block_devices[device_name]["PKNAME"]
+            else:
+                # find next ancestors from next parent_name
```

Comment:
If its not partition or disk, then what value for `block_devices[device_name]["TYPE"]`?

Reply:
it might be lvm, mpath or something else but we only need the find disks and partitions of block device which have a mount point. it can be any type, we need to find an ancestor of that block device. if that block device also a disk or partitions then consider that block device also its ancestor.

Reply:
ok

Reply:
I need all this information in the code as comments